text/3735-field-projections.md

```diff
@@ -0,0 +1,1344 @@
+                Poll::Pending => self->fut2.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        } else {
+            // self->fut2: Pin<&mut F2>
+            match self->fut2.poll(ctx) {
+                Poll::Pending => self->fut1.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        }
+    }
+}
+```
+
+## Impact of this Feature
+
+Overall this feature improves readability of code, because it replaces more complex to parse syntax
+with simpler syntax:
+- `&raw mut (*ptr).foo` is turned into `ptr->foo`
+- using `NonNull<T>` as a replacement for `*mut T` becomes a lot better when accessing fields,
+
+There is of course a cost associated with introducing a new operator along with the concept of field
+projections.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Implementation Details
+
+In order to facilitate field projections, several interlinked concepts have to be introduced. These
+concepts are:
+
+- [field types],
+  - `Field` trait,
+  - `field_of!` macro,
+  - [`#[projecting]`](#projecting-attribute) attribute
+- projection operator `->`,
+  - `Project` trait,
+  - `Projectable` trait,
+
+To ease understanding, here is a short explanation of the interactions between these concepts. The
+following subsections explain them in more detail, so refer to them in cases of ambiguity. The
+projection operator `->` is governed by the `Project` trait that has `Projectable` as a super trait.
+`Projectable` helps to select the struct whose fields are used for projection. Field types store
+information about a field (such as the base struct and the field type) via the `Field` trait and the
+`field_of!` macro makes it possible to name the [field types]. Finally, the
+[`#[projecting]`](#projecting-attribute) attribute allows `repr(transparent)` structs to be ignored
+when looking for fields for projection.
+
+### Field Types
+[field type]: #field-types
+[field types]: #field-types
+
+The compiler generates a compiler-internal type for every field of every struct. These types can
+only be named via the `field_of!` macro that has the same syntax as `offset_of!`. Only fields
+accessible to the current scope can be projected. These types are called *field types*.
+
+Field types implement the `Field` trait:
+
+```rust
+pub trait Field {
+    type Base: ?Sized;
+    type Type: ?Sized;
+
+    const OFFSET: usize;
+}
+```
+
+In the implementation of this trait, `Base` is set to the struct that the field is part of and
+`Type` is set to the type of the field. `OFFSET` is set to the offset in bytes of the field in
+the struct (i.e. `OFFSET = offset_of!(Base, ident)`).
+
+In addition to all fields of all structs, field types are generated for transparent,
+[`#[projecting]`](#projecting-attribute) container types as follows: given a transparent, generic type annotated with
+[`#[projecting]`](#projecting-attribute) and a struct contained in it:
```

Comment:
The requirement that the container be transparent prevents useful usecases like ["projected Arc"](https://docs.rs/pared/latest/pared/sync/struct.Parc.html). I think you can relax the requirement to something like the DST requirement, i.e. "the last field must be projectable for `T`".